cartographer/mapping_2d/local_trajectory_builder.h

```diff
@@ -41,10 +41,23 @@ namespace mapping_2d {
 // TODO(gaschler): Add test for this class similar to the 3D test.
 class LocalTrajectoryBuilder {
  public:
+  // The following two structs are used for representing the result of adding
+  // range data in local SLAM (using 'LocalTrajectoryBuilder::AddRangeData()').
+  // A null pointer is returned for 'MatchingResult' in case if the range data
+  // accumulation is not yet complete. Otherwise, a pointer is returned to a
+  // 'MatchingResult' instance which contains the accumulated 'RangeData',
+  // constant node data (although the node containing this data may not exist
+  // if the motion filter blocked insertion), and an 'InsertionResult' pointer,
+  // which is null if the motion filter has blocked the accumulated range data
+  // from insertion; otherwise, it points to an 'InsertionResult' instance.
   struct InsertionResult {
-    std::shared_ptr<const mapping::TrajectoryNode::Data> constant_data;
     std::vector<std::shared_ptr<const Submap>> insertion_submaps;
   };
+  struct MatchingResult {
+    std::shared_ptr<const sensor::RangeData> range_data;
```

Comment:
Why the pointer? Seems unnecessary to me.

Reply:
I wanted to ensure that unnecessary copying is avoided when transferring ownership of `range_data` to the library user. Although a unique pointer would be more appropriate since `range_data` is not used in rest of Cartographer, unlike `constant_data`. You think that we don't need pointers at all? I have always felt more comfortable when handling structures containing lots of data using pointers rather than relying on RVO or using move semantics.

Reply:
Move semantics should achieve what you want. Otherwise: Do not optimize prematurely.